lib/web/websocket/stream/websocketstream.js

```diff
@@ -0,0 +1,494 @@
+      // 8.1. Let signal be options [" signal "].
+      const signal = options.signal
+
+      // 8.2. If signal is aborted , then reject this 's opened promise and closed promise with signal ’s abort reason
+      //      and return.
+      if (signal.aborted) {
+        this.#openedPromise.reject(signal.reason)
+        this.#closedPromise.reject(signal.reason)
+        return
+      }
+
+      // 8.3. Add the following abort steps to signal :
+      signal.addEventListener('abort', () => {
+        // 8.3.1. If the WebSocket connection is not yet established : [WSP]
+        if (!isEstablished(this.#handler.readyState)) {
+          // 8.3.1.1. Fail the WebSocket connection .
+          failWebsocketConnection(this.#handler)
+
+          // Set this 's ready state to CLOSING .
+          this.#handler.readyState = states.CLOSING
+
+          // Reject this 's opened promise and closed promise with signal ’s abort reason .
+          this.#openedPromise.reject(signal.reason)
+          this.#closedPromise.reject(signal.reason)
+
+          // Set this 's handshake aborted to true.
+          this.#handshakeAborted = true
+        }
+      }, { once: true })
+    }
+
+    // 9.  Let client be this 's relevant settings object .
+    const client = environmentSettingsObject.settingsObject
+
+    // 10. Run this step in parallel :
+    // 10.1. Establish a WebSocket connection given urlRecord , protocols , and client . [FETCH]
+    this.#controller = establishWebSocketConnection(
+      urlRecord,
+      protocols,
+      client,
+      this.#handler,
+      options
+    )
+  }
+
+  // The url getter steps are to return this 's url , serialized .
+  get url () {
+    return this.#url.toString()
+  }
+
+  // The opened getter steps are to return this 's opened promise .
+  get opened () {
+    return this.#openedPromise.promise
+  }
+
+  // The closed getter steps are to return this 's closed promise .
+  get closed () {
+    return this.#closedPromise.promise
+  }
+
+  // The close( closeInfo ) method steps are:
+  close (closeInfo = undefined) {
+    if (closeInfo !== null) {
+      closeInfo = webidl.converters.WebSocketCloseInfo(closeInfo)
+    }
+
+    // 1. Let code be closeInfo [" closeCode "] if present, or null otherwise.
+    const code = closeInfo.closeCode ?? null
+
+    // 2. Let reason be closeInfo [" reason "].
+    const reason = closeInfo.reason
+
+    // 3. Close the WebSocket with this , code , and reason .
+    closeWebSocketConnection(this.#handler, code, reason, true)
+  }
+
+  #write (chunk) {
+    // 1. Let promise be a new promise created in stream ’s relevant realm .
+    const promise = createDeferredPromise()
+
+    // 2. Let data be null.
+    let data = null
+
+    // 3. Let opcode be null.
+    let opcode = null
+
+    // 4. If chunk is a BufferSource ,
+    if (ArrayBuffer.isView(chunk) || types.isAnyArrayBuffer(chunk)) {
+      // 4.1. Set data to a copy of the bytes given chunk .
+      data = Buffer.from(chunk)
```

Comment:
```suggestion
      data = new Uint8Array(ArrayBuffer.isView(chunk) ? new Uint8Array(chunk.buffer, chunk.byteOffset, chunk.byteLength) : chunk)
```

Handling of non-Uint8Array